addon/services/indexed-db.js

```diff
@@ -609,26 +610,36 @@ export default class IndexedDbService extends Service {
       // If a multi index is found, use it
       if (multiIndex) {
         let keyPath = get(multiIndex, 'keyPath');
-        let compareValues = array();
-        keyPath.forEach((key) => {
-          compareValues.push(query[key]);
-        });
+        let compareValues = keyPath.map((key) => query[key]);
 
         let keyName = get(multiIndex, 'name');
         return db[type].where(keyName).equals(compareValues);
       }
     }
 
     // Else, filter manually
-    Object.keys(query).forEach((i) => {
-      if (!promise) {
-        promise = db[type].where(i).equals(query[i]);
-      } else {
-        promise = promise.and((item) => get(item, i) === query[i]);
-      }
-    });
-
-    return promise;
+    return Object.entries(query)
+      .reduce(
+        (query, [queryKey, queryValue], i) => {
+          let index = indexMap[queryKey];
+
+          assert(index, `You are attempting to query ${queryKey} which is not a valid index for ${type}`);
+
+          let isMulti = index.multi && isArray(queryValue);
+          let first = i === 0;
+
+          if (first) {
+            return isMulti
+              ? query.where(queryKey).anyOf(...queryValue)
+              : query.where(queryKey).equals(queryValue);
+          } else {
+            let predicate = isMulti
+              ? (item) => get(item, queryKey).any(_ => queryValue.includes(_))
+              : (item) => get(item, queryKey) === queryValue;
+
+            return query.and(predicate);
+          }
+        }, db[type]);
   }
```

Comment:
We should probably add `distinct()` before we return it, if at least one is a multi entry index (see: https://dexie.org/docs/MultiEntry-Index)

Reply:
Yeah, that makes sense, I was also thinking we should maybe sort the indexes by multi beforehand, from a performance perspective I would expect that having IndexedDb do at least one ```#anyOf``` query would get you some benefit overdoing the set intersection in memory.